deep_quoridor/src/arena.py

```diff
@@ -148,53 +189,59 @@ def _play_game(self, agent1: Agent, agent2: Agent, game_id: str) -> GameResult:
-            total_games = (len(agents) - 1) * times
+            first_agent = players[0]
+            remaining_agents = players[1:]
 
-        self.plugins.start_arena(self.game, total_games=total_games)
+            for opp in remaining_agents:
+                for _ in range(times):
+                    opponent = random.choice(remaining_agents) if mode == PlayMode.FIRST_VS_RANDOM else opp
+                    if not self.swap_players or match_id % 2 == 0:
+                        games_to_play.append((first_agent, opponent, f"game_{match_id:04d}"))
+                    else:
+                        games_to_play.append((opponent, first_agent, f"game_{match_id:04d}"))
+
+                    match_id += 1
+
+        return games_to_play
+
+    def _play_games(self, players: list[str], times: int, mode: PlayMode, num_workers: int = 0) -> list[GameResult]:
+        games_to_play = self._list_of_games_to_play(players, times, mode)
+        self.plugins.start_arena(self.game, total_games=len(games_to_play))
 
-        match_id = 1
         results = []
 
         try:
-            if mode == PlayMode.ALL_VS_ALL:
-                for i in range(len(agents)):
-                    for j in range(i + 1, len(agents)):
-                        for t in range(times):
-                            agent_1, agent_2 = (
-                                (agents[i], agents[j])
-                                if not self.swap_players or t % 2 == 0
-                                else (agents[j], agents[i])
-                            )
-                            result = self._play_game(agent_1, agent_2, f"game_{match_id:04d}")
-                            results.append(result)
-                            match_id += 1
-            else:  # FIRST_VS_RANDOM or FIRST_VS_ALL mode
-                first_agent = agents[0]
-                remaining_agents = agents[1:]
-
-                for opp in remaining_agents:
-                    for _ in range(times):
-                        opponent = random.choice(remaining_agents) if mode == PlayMode.FIRST_VS_RANDOM else opp
-                        agent_1, agent_2 = (
-                            (first_agent, opponent)
-                            if not self.swap_players or match_id % 2 == 0
-                            else (opponent, first_agent)
-                        )
-                        result = self._play_game(agent_1, agent_2, f"game_{match_id:04d}")
+            if num_workers == 0:
+                # play in this process
+                agents = {}
+                for p in players:
+                    agents[p] = AgentRegistry.create_from_encoded_name(p, self.game)
+
+                for p1, p2, game_id in games_to_play:
+                    result = self._play_game(agents[p1], agents[p2], game_id)
+                    results.append(result)
+            else:
+                # Multi process play
+                with concurrent.futures.ProcessPoolExecutor(
```

Comment:
I haven't used the ProcessPoolExecutor before - I like how simple it makes the code.

Reply:
Yeah, it saves you from having to explicitly use queues